src/psyclone/tests/domain/lfric/transformations/lfric_extract_test.py

```diff
@@ -0,0 +1,698 @@
+        "m1_proxy%data, m2_proxy%data, ndf_w1, undf_w1, " + \
+        "map_w1(:,cell), ndf_w2, undf_w2, map_w2(:,cell), ndf_w3, " + \
+        """undf_w3, map_w3(:,cell))
+      END DO
+      CALL psy_data%PostStart
+      CALL psy_data%ProvideVariable("cell_post", cell)
+      CALL psy_data%ProvideVariable("f1_post", f1)
+      CALL psy_data%PostEnd
+      !
+      ! ExtractEnd"""
+    assert output in code
+
+    # TODO #646
+    # At this stage not all required parameters are passed via PSyData.
+    # This is an except of missing lines, which will cause this test to x-fail
+    not_yet_working = ['CALL psy_data%ProvideVariable("nlayers", nlayers)',
+                       'CALL psy_data%ProvideVariable("m1_proxy", m1_proxy)',
+                       'CALL psy_data%ProvideVariable("ndf_w1", ndf_w1)',
+                       'CALL psy_data%ProvideVariable("undf_w1", undf_w1)']
+    for line in not_yet_working:
+        if line not in code:
+            pytest.xfail("#646 LFRic extraction not fully supported yet.")
+    assert False, "X-failing test suddenly working: #646 LFRic extraction."
+
+
+def test_node_list_dynamo0p3():
+    ''' Test that applying Extract Transformation on a list of Nodes
+    produces the correct result in Dynamo0.3 API. '''
+    etrans = LFRicExtractTrans()
+    psy, invoke = get_invoke("15.1.2_builtin_and_normal_kernel_invoke.f90",
+                             DYNAMO_API, idx=0, dist_mem=False)
+    schedule = invoke.schedule
+
+    schedule, _ = etrans.apply(schedule.children[0:3])
+    code = str(psy.gen)
+    # This output is affected by #637 (builtin support), and needs to be
+    # adjusted once this is fixed.
+    output = """! ExtractStart
+      !
+      CALL psy_data%PreStart("single_invoke_builtin_then_kernel_psy", """ \
+      """"invoke_0:r0", 1, 3)
+      CALL psy_data%PreDeclareVariable("f2", f2)
+      CALL psy_data%PreDeclareVariable("cell_post", cell)
+      CALL psy_data%PreDeclareVariable("df_post", df)
+      CALL psy_data%PreDeclareVariable("f3_post", f3)
+      CALL psy_data%PreEndDeclaration
+      CALL psy_data%ProvideVariable("f2", f2)
+      CALL psy_data%PreEnd
+      DO df=1,undf_any_space_1_f5
+        f5_proxy%data(df) = 0.0
+      END DO
+      DO df=1,undf_any_space_1_f2
+        f2_proxy%data(df) = 0.0
+      END DO
+      DO cell=1,f3_proxy%vspace%get_ncell()
+        !
+        CALL testkern_code_w2_only(nlayers, f3_proxy%data, """ + \
+        """f2_proxy%data, ndf_w2, undf_w2, map_w2(:,cell))
+      END DO
+      CALL psy_data%PostStart
+      CALL psy_data%ProvideVariable("cell_post", cell)
+      CALL psy_data%ProvideVariable("df_post", df)
+      CALL psy_data%ProvideVariable("f3_post", f3)
+      CALL psy_data%PostEnd
+      !
+      ! ExtractEnd"""
+    assert output in code
+
+    # TODO #646
+    # At this stage not all required parameters are passed via PSyData.
+    # This is an except of missing lines, which will cause this test to x-fail
```

Comment:
s/except/excerpt/

Reply:
Now here is finally the business idea that will make me rich: I solve the copy&paste problem.